src/Compiler/Driver/CompilerDiagnostics.fs

```diff
@@ -617,12 +617,28 @@ let OutputNameSuggestions (os: StringBuilder) suggestNames suggestionsF idText =
                     os.AppendString "   "
                     os.AppendString(ConvertValLogicalNameToDisplayNameCore value)
 
+let OutputTypesNotInEqualityRelationContextInfo contextInfo ty1 ty2 m (os: StringBuilder) fallback =
+    match contextInfo with
+    | ContextInfo.IfExpression range when equals range m -> os.AppendString(FSComp.SR.ifExpression (ty1, ty2))
+    | ContextInfo.CollectionElement (isArray, range) when equals range m ->
+        if isArray then
+            os.AppendString(FSComp.SR.arrayElementHasWrongType (ty1, ty2))
+        else
+            os.AppendString(FSComp.SR.listElementHasWrongType (ty1, ty2))
+    | ContextInfo.OmittedElseBranch range when equals range m -> os.AppendString(FSComp.SR.missingElseBranch (ty2))
+    | ContextInfo.ElseBranchResult range when equals range m -> os.AppendString(FSComp.SR.elseBranchHasWrongType (ty1, ty2))
+    | ContextInfo.FollowingPatternMatchClause range when equals range m ->
+        os.AppendString(FSComp.SR.followingPatternMatchClauseHasWrongType (ty1, ty2))
+    | ContextInfo.PatternMatchGuard range when equals range m -> os.AppendString(FSComp.SR.patternMatchGuardIsNotBool (ty2))
+    | contextInfo -> fallback contextInfo
+
 type Exception with
 
     member exn.Output(os: StringBuilder, suggestNames) =
 
         match exn with
-        | ConstraintSolverTupleDiffLengths (_, tl1, tl2, m, m2) ->
+        // TODO: this is now unused...?
+        | ConstraintSolverTupleDiffLengths (_, _, tl1, tl2, m, m2) ->
             os.AppendString(ConstraintSolverTupleDiffLengthsE().Format tl1.Length tl2.Length)
 
             if m.StartLine <> m2.StartLine then
```

Comment:
I guess there's still this part, which I think can now be removed, including the associated error message.

Reply:
Although I can't say with 100% certainty that this exception will never end up here. So maybe it's safer to keep it for now.

Reply:
=> Keep